src/main/java/mate/academy/dao/BookDaoImpl.java

```diff
@@ -0,0 +1,148 @@
+            throw new DataProcessingException(CREATE_OPERATION_PERFORMING_MESSAGE
+                    + book.getTitle()
+                    + " "
+                    + book.getPrice(), e);
+        }
+        return book;
+    }
+
+    @Override
+    public Optional<Book> findById(Long id) {
+        Book book = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    FIND_BY_ID_QUERY);
+            statement.setLong(FIRST_QUERY_PARAM_INDEX, id);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                book = getBookFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException(CANT_FIND_BOOK_MESSAGE + id, e);
+        }
+        return Optional.ofNullable(book);
+    }
+
+    @Override
+    public List<Book> findAll() {
+        List<Book> bookList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    FIND_ALL_QUERY);
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                bookList.add(getBookFromResultSet(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException(CANT_GET_ALL_BOOKS_MESSAGE, e);
+        }
+        return bookList;
+    }
+
+    @Override
+    public Book update(Book book) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(UPDATE_BY_ID_QUERY);
+            statement.setString(FIRST_QUERY_PARAM_INDEX, book.getTitle());
+            statement.setObject(SECOND_QUERY_PARAM_INDEX, book.getPrice());
+            statement.setLong(THIRD_QUERY_PARAM_INDEX, book.getId());
+            int affectedRows = statement.executeUpdate();
+            checkIfAtLeastOneRowAffected(affectedRows);
+            Optional<Book> updatedBook = findById(book.getId());
+            return updatedBook.get();
+        } catch (SQLException e) {
+            throw new DataProcessingException(CANT_UPDATE_BOOK_MESSAGE + book.getId(), e);
+        }
+    }
+
+    @Override
+    public boolean deleteById(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(DELETE_BY_ID_QUERY);
+            statement.setLong(FIRST_QUERY_PARAM_INDEX, id);
+            int affectedRows = statement.executeUpdate();
+            checkIfAtLeastOneRowAffected(affectedRows);
+            // Mb, we need to return statement like this: affectedRows == 1, because,
+            // it's strange, if we delete several rows with one id (but we don't have guarantee,
+            // that id unique in random DB), so what?
```

Comment:
I think that's fine for now, we just want to make sure that rows are affected here. And you are right, for some reasons we can have id field not unique, so probably in some cases that would be fine, although that's rather exceptional situation

Reply:
Thanks, the comments were deleted.